avahi-daemon/static-services.c

```diff
@@ -475,9 +619,46 @@ static void XMLCALL xml_end(void *data, AVAHI_GCC_UNUSED const char *el) {
 
         case XML_TAG_TXT_RECORD: {
             assert(u->service);
+            if (u->txt_key != NULL) {
+                size_t key_len = strlen(u->txt_key);
+                uint8_t *value_buf;
+                uint8_t *free_value_buf = NULL;
+                size_t value_buf_len = 0;
+
+                switch (u->txt_type) {
+                    case TXT_RECORD_VALUE_TEXT:
+                        value_buf_len = strlen(u->buf);
+                        value_buf = (uint8_t*)u->buf;
+                        break;
+
+                    case TXT_RECORD_VALUE_BINARY_HEX:
+                        if (decode_hex_buf (u, &value_buf, &value_buf_len) < 0)
+                            return;
+                        free_value_buf = value_buf;
+                        break;
+
+                    case TXT_RECORD_VALUE_BINARY_BASE64:
+                        if (decode_base64_buf (u, &value_buf, &value_buf_len) < 0)
+                            return;
+                        free_value_buf = value_buf;
+                        break;
+
+                    default:
+                        assert(0);
+                }
+
+                u->service->txt_records = avahi_string_list_add_anonymous(u->service->txt_records, key_len + 1 + value_buf_len);
```

Comment:
Hm, actually, looking at the `avahi_string_list_add_anonymous`, it looks like it allocates a buffer 1 byte larger than requested and terminates the buffer with nul for us.